include/boost/windows_sspi/stream.hpp

```diff
@@ -275,44 +257,22 @@ template <typename NextLayer> class stream : public stream_base {
 
   template <typename MutableBufferSequence>
   size_t read_some(const MutableBufferSequence& buffers, boost::system::error_code& ec) {
-    SECURITY_STATUS sc = SEC_E_OK;
-    while(m_sspi_impl->decrypt.decrypted_data.empty()) {
-      // TODO: This is duplicated in the async version. Move logic to
-      // sspi_impl some way. Possibly a decryption class for keep
-      // track of state and buffers.
-      std::vector<char> input_buffer;
-      if (m_sspi_impl->decrypt.encrypted_data.empty() || sc == SEC_E_INCOMPLETE_MESSAGE) {
-        input_buffer.resize(0x10000);
-        std::size_t size_read = m_next_layer.read_some(net::buffer(input_buffer.data(), input_buffer.size()), ec);
-        if (ec && size_read == 0 && m_sspi_impl->decrypt.decrypted_data.empty()) {
-          return 0;
-        }
-        input_buffer.resize(size_read);
-      }
-      sc = m_sspi_impl->decrypt(input_buffer);
-      input_buffer.clear();
-      if (sc == SEC_E_INCOMPLETE_MESSAGE) {
-        continue;
-      }
-      if (sc == SEC_I_CONTEXT_EXPIRED) {
-        // TODO: Shutdown the TLS context gracefully (implement
-        // shutdown), then return EOF.
-        ec = net::error::eof;
-        return 0;
-      }
-      if (sc != SEC_E_OK) {
-        ec = error::make_error_code(sc);
-        return 0;
-      }
-      break;
+    while(m_sspi_impl->decrypt() == detail::sspi_decrypt::state::data_needed) {
+      std::array<char, 0x10000> input_buffer;
+      std::size_t size_read = m_next_layer.read_some(net::buffer(input_buffer.data(), input_buffer.size()), ec);
+      m_sspi_impl->decrypt.put({input_buffer.begin(), input_buffer.begin() + size_read});
+      continue;
+    }
+
+    if (m_sspi_impl->decrypt() == detail::sspi_decrypt::state::error) {
+      ec = boost::error::make_error_code(m_sspi_impl->decrypt.error_code);
```

Comment:
`make_error_code` is designed to be found through ADL. This doesn't work for you because `SECURITY_STATUS` is not an enum.
If it were, you could create a specialilsation of `boost::system::is_error_code_enum<SECURITY_STATUS> : boost::true_type {};` and then simply write `ec = m_sspi_impl->decrypt.error_code;`

Reply:
I've been struggling a bit with making this work. Mostly because (I think) `SECURITY_STATUS` is actually just a `long` type, so I either the conversion to `error_code` already exists, or there's no conversion to a custom enum type if I declare that.

I'm sure I'll figure that out though, there are quite a few other boost libraries where I can find inspiration.

Reply:
the trick is to make it an enum and provide the boost::system plumbing (error_category, specialisation of is_enum_error_code<> and ADL make_error_code). If it's just copying values from the C api the enum could be empty. Then assignment becomes:
```
ec = decrypt.error_code;
```

Reply:
Asio and beast have examples (boost/asio/error.hpp)

Reply:
Thanks. I was a bit confused as to whether I could simply make it work without having to redeclare all the SECURITY_STATUS errors as enum values. Looking at the asio code, it does indeed seems like it possible, so thanks for clarifying that.

I'll look into that, but I'll probably just merge this PR as it is.